runtime/buffer.h

```diff
@@ -0,0 +1,179 @@
+  std::size_t FrameLength() const {
+    return std::min(length_ - frame_, size_ - (start_ + frame_));
+  }
+
+  // Returns a short frame at a non-fatal EOF.  Can return a long frame as well.
+  std::size_t ReadFrame(
+      FileOffset at, std::size_t bytes, IoErrorHandler &handler) {
+    Flush(handler);
+    Reallocate(bytes, handler);
+    if (at < fileOffset_ || at > fileOffset_ + length_) {
+      Reset(at);
+    }
+    frame_ = static_cast<std::size_t>(at - fileOffset_);
+    if (start_ + frame_ + bytes > size_) {
+      DiscardLeadingBytes(frame_, handler);
+      if (start_ + bytes > size_) {
+        // Frame would wrap around; shift current data (if any) to force
+        // contiguity.
+        RUNTIME_CHECK(handler, length_ < size_);
+        if (start_ + length_ <= size_) {
+          // [......abcde..] -> [abcde........]
+          std::memmove(buffer_, buffer_ + start_, length_);
+        } else {
+          // [cde........ab] -> [abcde........]
+          auto n{start_ + length_ - size_};  // 3 for cde
+          auto gap{size_ - length_};  // 13 - 5 = 8
+          RUNTIME_CHECK(handler, length_ >= n);
+          std::memmove(buffer_ + n, buffer_ + start_, length_ - n);  // cdeab
+          LeftShiftBufferCircularly(buffer_, length_, n);  // abcde
+        }
+        start_ = 0;
+      }
+    }
+    while (FrameLength() < bytes) {
+      auto next{start_ + length_};
+      RUNTIME_CHECK(handler, next < size_);
+      auto chunk{size_ - next};
+      auto got{
+          Store().Read(fileOffset_ + length_, buffer_ + next, chunk, handler)};
+      length_ += got;
+      RUNTIME_CHECK(handler, length_ < size_);
+      if (got < chunk) {
+        break;  // error or EOF & program can handle it
+      }
+    }
+    return FrameLength();
+  }
+
+  void WriteFrame(FileOffset at, std::size_t bytes, IoErrorHandler &handler) {
+    if (!dirty_ || at < fileOffset_ || at > fileOffset_ + length_ ||
+        start_ + (at - fileOffset_) + bytes > size_) {
+      Flush(handler);
+      fileOffset_ = at;
+      Reallocate(bytes, handler);
+    }
+    dirty_ = true;
+    frame_ = at - fileOffset_;
+    length_ = std::max(length_, frame_ + bytes);
+  }
+
+  bool Flush(IoErrorHandler &handler) {
+    if (dirty_) {
+      while (length_ > 0) {
+        std::size_t chunk{std::min(length_, size_ - start_)};
+        std::size_t put{
+            Store().Write(fileOffset_, buffer_ + start_, chunk, handler)};
+        length_ -= put;
+        start_ += put;
+        fileOffset_ += put;
+        if (put < chunk) {
+          break;
+        }
+      }
+      Reset(fileOffset_);
+    }
+  }
+
+private:
+  STORE &Store() { return static_cast<STORE &>(*this); }
+
+  void Reallocate(std::size_t bytes, Terminator &terminator) {
+    if (bytes > size_) {
+      char *old{buffer_};
+      auto oldSize{size_};
+      size_ = std::max(bytes, minBuffer);
+      buffer_ = AllocateMemoryOrCrash(size_, terminator);
```

Comment:
GNU 8.3.0 gives an error on this line:
../runtime/buffer.h:134:39: error: invalid initialization of reference of type 'Fortran::runtime::Terminator&' from expression of type 'std::size_t' {aka 'long unsigned int'}

Reply:
Fixed.